src/selectors.js

```diff
@@ -72,31 +92,102 @@ const _buildRecords = (ents, key, opts) => {
   return finalRecords;
 };
 
+const makeNestedJoins = ({ children, records, entities }) => {
+  return children.reduce((acc, child) => {
+    const parentRecords = acc.length ? acc : records;
+    const childRecords = enrichRecords(entities, child.key);
+    return joins({
+      items: parentRecords,
+      mergeItems: childRecords,
+      tableName: child.parent,
+      mergeTableName: child.key,
+    });
+  }, []);
+};
+
+const buildNestedFullRecords = ({ joinKey, children, nestedJoins }) => {
+  return nestedJoins.map((record) =>
+    buildFullRecord(
+      record,
+      joinKey,
+      children.map((c) => c.key)
+    )
+  );
+};
+
 const defaultArr = [];
 
-const buildRecords = (ents = {}, key, opts = {}) => {
-  // console.log(`All. key: ${key}`, ents, opts)
-  if (!_.isObject(opts)) opts = {};
-  const { model, defaultJoins } = getModel(key);
-  const joinKeys = utils.composeArray(opts.joins, defaultJoins);
-  let final = enrichRecords(ents, key) || defaultArr;
-
-  joinKeys.forEach((k) => {
-    const mergeItems = enrichRecords(ents, k);
-    final = joins({
-      items: final,
-      mergeItems,
+const joinsMapper = ({ joinKeys, entities, key }) => {
+  let finalRecords = enrichRecords(entities, key) || defaultArr;
+
+  joinKeys.forEach(({ key: joinKey, children }) => {
+    const records = enrichRecords(entities, joinKey);
+    const nestedJoins = makeNestedJoins({ children, records, entities });
+    const fullRecords = buildNestedFullRecords({
+      joinKey,
+      children,
+      nestedJoins,
+    });
+
+    const hasJoinsThrough = fullRecords.length;
+
+    finalRecords = joins({
+      items: finalRecords,
+      mergeItems: hasJoinsThrough
+        ? fullRecords
+        : enrichRecords(entities, joinKey),
       tableName: key,
-      mergeTableName: k,
+      mergeTableName: joinKey,
     });
   });
 
-  const assoc = config.createAssociation(
-    final.map((i) => buildFullRecord(i, key, joinKeys))
+  return finalRecords;
+};
+
+const buildRecords = (entities = {}, key, opts = {}) => {
+  opts = _.isObject(opts) ? opts : {};
+  const { defaultJoins = [] } = getModel(key) || {};
+  const joinsArray = utils.composeArray(defaultJoins, opts.joins);
+  const joinKeys = buildNestedStructure(joinsArray, key);
+  const finalRecords = joinsMapper({ joinKeys, entities, key });
+
+  const associatedRecords = config.createAssociation(
+    finalRecords.map((record) =>
+      buildFullRecord(
+        record,
+        key,
+        joinKeys.map(({ key }) => key)
+      )
+    )
   );
-  return assoc;
+
+  return associatedRecords;
 };
 
+function buildNestedStructure(input, parentKey = null) {
+  let nodes = [];
+
+  if (_.isString(input)) {
+    return [{ key: input, parent: parentKey, children: [] }];
+  } else if (_.isArray(input)) {
+    input.forEach((item) => {
+      nodes = nodes.concat(buildNestedStructure(item, parentKey));
+    });
+  } else if (_.isObject(input)) {
+    _.forOwn(input, (value, key) => {
```

Comment:
waarom gebruik je hier `forOwn` en niet `_.map`?

Reply:
ipv forEach bedoel je? omdat het hier een object is en geen array en ik zo zowel de key als value beschikbaar heb